packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -546,31 +581,42 @@ class _TextSelectionHandleOverlayState extends State<_TextSelectionHandleOverlay
         break;
     }
 
+    final Size viewport = widget.renderObject.size;
+    if (!viewport.contains(point)) {
+      point = Offset(
+        point.dx.clamp(0.0, viewport.width),
+        point.dy.clamp(0.0, viewport.height),
+      );
+    }
```

Comment:
Seems like it'd be cheaper to just always clamp?

Reply:
I considered that, but I imagined that the "common case" is where both handles are visible, in which case the clamping doesn't change the input offset. It seemed better to me to avoid unnecessary allocation in this case.

Reply:
It would also be easier to read if we always clamp and allocations are supposed to be cheap in dart... If you have a benchmark showing that this is better, we can keep it. Otherwise I'd suggest changing it to the simplified version.

Reply:
Done.